proposals/004-health-checks.md

```diff
@@ -0,0 +1,87 @@
+runs as a background thread and orchestrates starting the routes. The `RouteController` will then be responsbible for starting up all these routes, and have support for periodically retry in case a route fails to startup. And only when all routes has been started the `RouteController` has completed starting up all the routes.
+
+We would need to make the discovery and bootstrap of the `RouteController` very easy, which we can do using Camel's existing `RoutePolicyFactory`.
+
+In Camel with Spring Boot we can setup this in the `Main` class via something along the following code lines:
+
+    @Bean
+    RouteController newRouteController(CamelContext camelContext) {
+        return new RouteController(camelContext);
+    }
+
+    @Bean
+    RoutePolicyFactory newRouteController(RouteController controller) {
+        return new RouteControllerRoutePolicyFactory(controller);
+    }
+
+Or:
+
+    @Bean
+    RouteController newRouteController(CamelContext camelContext) {
+        // The route controller is also in charge to add add its own
+        // RoutePolicyFactory to the camel context
+        return new RouteController(camelContext);
+    }
+
+There should be one shared instance of `RouteController` which is given to the `RoutePolicyFactory`, which then creates a `RoutePolicy` for each route that will be controlled.
+
+TODO: We could make `RouteController` an SPI in Apache Camel and provide it OOTB in Apache Camel. Then we can have Camel auto detect if its enabled on startup
+and then automatic set `autoStartup = false`, and then let the `RouteController` startup the routes (We have this idea for Camel 3.0). This can then make it even
+easier to enable. The `RouteController` should have configuration setting to specify how often to retry starting failed routes (we may need to have backoff etc).
+We can then implement a `DefaultRouteController` in camel-core, and then make it easy to use by setting `CamelContext.setRouteControllerEnabled(true)` (find a good name). With Camel on Spring Boot then its a matter of setting this in the `application.properties` via `camel.springboot.route-controller.enabled = true`.
+
+Besides starting up routes the `RouteController` should have a Java, JMX API, and REST which can report back status of the running integrations (i.e. running routes).
+At minimum it should collect status of each route with details:
+
+- route id
+- current status (starting, started, stopping, stopped)
+- health (unknown, ok, failure)
+- failure error message (exception message + stacktrace)
+
+TODO: API details to be determined
+
+To expose this API in Rest we can use Spring Actuate which allows to expose such details. For en example see `org.apache.camel.spring.boot.actuate.endpoint.CamelRoutesEndpoint`
+
+The `RouteController` should also have APIs to force starting a route on demand, eg so an user can click a button, and then it will immediately schedule to run a task that attempts to start the route.
```

Comment:
That probably answers my quesstion above. But its not about _changing_ the route, just starting and stopping, right ?

For changing a route (including configuration options) I'd still prefer doing a full redeployment (stopping old integration pod, creating and starting a new one based on a new branch in git with the new version) in order to keep it simple and keeping a clear deployment model with immutable containers (but even when keeping the pod one would have to restart the whole app when the route changes which makes it not much different as recreating the whole pod ?)